### ISSUE TITLE
--config-command="..." or --config-key-command="..." option

### ISSUE THREAD

**Issue by EduardDurech** [rclone/rclone#6447]
Version: rclone v1.59.1

#### What problem are you are trying to solve?

It would be nice to have a `--config-command` as analogues to the `--password-command` so we can source our own encryption calls for the config file, or provide our own master encryption key (**not** password) to supplement in:

https://github.com/rclone/rclone/blob/f0e439de0df77b6ceee197d62f51b47455fe8a3a/fs/config/config.go#L80

https://github.com/rclone/rclone/blob/f0e439de0df77b6ceee197d62f51b47455fe8a3a/fs/config/config.go#L275

This is particularly useful when we have devices which have their own hardware-backed keystore, such as smartphones. I am [currently](https://github.com/termux/termux-api/issues/550) implementing in Termux the ability to encrypt and decrypt using AES in Android's keystore. The problem with the current method is that I would still need to include the encrypted master password somewhere (such as in a file or sharedpreferences), but it would be preferable if I could source the master key directly from the keystore, or call a command to provide the decrypted config, therefore making sure the password isn't stored anywhere feasibly accessible on the device (even in encrypted form), nor in memory/cache.

Android's keystore supports biometric and pattern authentication and can be unlocked for a certain duration, only when the device is unlocked, for x times, et cetera, so the security of the config file would be greater and never expose the key (encrypted or unencrypted) outside the call.

It would be something like [connection strings](https://rclone.org/docs/#connection-strings), having no persistent configuration data, but maintaining the convenience of not having to manually input the configuration details

**Comment by ncw** [comment]
> It would be nice to have a `--config-command` as analogues to the `--password-command` so we can source our own encryption calls for the config file

There is a pluggable interface for the config file if you are willing to write some Go code

https://github.com/rclone/rclone/blob/1c99661d8cfbafa86ea186a3d0c8f07993a122a9/fs/config/config.go#L69

> or provide our own master encryption key (**not** password) to supplement in:

I'm not following you here - what is the difference between a key and a password other than one might be user readable and the other not?

> The problem with the current method is that I would still need to include the encrypted master password somewhere (such as in a file or sharedpreferences), but it would be preferable if I could source the master key directly from the keystore

Doesn't `--password-command ` let you do this? You write a command to return the password from the keystore and rclone uses it? Or am I mis-understanding? I don't know how the keystore works so maybe that is part of it.

> or call a command to provide the decrypted config, therefore making sure the password isn't stored anywhere feasibly accessible on the device (even in encrypted form), nor in memory/cache.

Remember rclone needs to be able to write to the config too (to store updated tokens) so we'd need an encryption and decryption interface.

**Comment by EduardDurech** [comment]
> There is a pluggable interface for the config file if you are willing to write some Go code

Yes, just currently working on the Termux keystore and other stuff, it may be some time until I could get around to it :/

> Doesn't `--password-command ` let you do this? You write a command to return the password from the keystore and rclone uses it? Or am I mis-understanding? I don't know how the keystore works so maybe that is part of it.

Yes but I'd like to return the unencrypted config file, not just the password. In Android keystore for example, the raw key is not accessible, but can be called via CLI to decrypt a supplied argument, such as the encrypted config file, and this could then be sourced instead of having to write unencrypted to storage then deleting after (for similar behaviour)

> I'm not following you here - what is the difference between a key and a password other than one might be user readable and the other not?

This was just a secondary thought that would be nice to have for possibly more use-cases, though upon further thought I think this could be emulated by creating the config via another script #3127 and then sourcing `--password-command`, but this was not my main intention as the keys in Android keystore are not directly accessible, so scratch that. I was just thinking about a way to encrypt and decrypt without CLI passwords being exposed by sourcing the key for encryption

> Remember rclone needs to be able to write to the config too (to store updated tokens) so we'd need an encryption and decryption interface.

Keystore has both encrypt and decrypt :) So we could supply a `--config-encrypt-command` and `--config-decrypt command`, or just a `--config-command` which the use would have to know which is the proper command he is calling in the use case.

In the case of Android (or Java applications), we would define a [`CipherInputStream`](https://developer.android.com/reference/javax/crypto/CipherInputStream) for encryption and a `CipherOutputStream` for decryption. This is what I am currently implementing in Termux's keystore

As of now, the way to emulate this would be to have a separate script decrypt some shell file which would send the arguments (such as file crypt passwords) in a connection string, but I feel this still exposes them to buffer or logs? Regardless, it would be nice to have it within rclone so don't have to have multiple scripts :)

**Comment by EduardDurech** [comment]
> There is a pluggable interface for the config file if you are willing to write some Go code

Hello @ncw, currently trying to do this, may I ask what IDE you are using? I am trying to use GoLand but it seems to be a problem debugging it as I can't debug (attach the process) before running the CLI (and, subsequently, breakpoints don't work)

A "hacky" way is adding a `bufio.NewReader`, but regardless, trying to see variables gives `can not call optimized function when regabi is in use`. I'm assuming there's a better way to do this

**Comment by ncw** [comment]
> Hello @ncw, currently trying to do this, may I ask what IDE you are using? I am trying to use GoLand but it seems to be a problem debugging it as I can't debug (attach the process) before running the CLI (and, subsequently, breakpoints don't work)
> 
> A "hacky" way is adding a `bufio.NewReader`, but regardless, trying to see variables gives `can not call optimized function when regabi is in use`. I'm assuming there's a better way to do this

Sorry can't help you there, I'm strictly a debug with `printf` kind of guy!

> Edit: Another question When entering config\config.go `init()`
> 
> https://github.com/rclone/rclone/blob/1c99661d8cfbafa86ea186a3d0c8f07993a122a9/fs/config/config.go#L116-L126
> 
> the call `fs.GetConfig(context.Background())` gives default values for trying to parse arguments. I'm assuming this is because `AddFlags(...)` and `SetFlags(...)` are called after `init()` What I would like to do is check if the new variable `ConfigCommandIn` was supplied so that
> https://github.com/rclone/rclone/blob/1c99661d8cfbafa86ea186a3d0c8f07993a122a9/fs/config/config.go#L123
> 
> isn't called. Is this possible with the current setup? Or would I have to call `os.Args` and cycle through them, unnecessarily parsing them twice
> Edit2: Just realized that
> 
> https://github.com/rclone/rclone/blob/966654e23a2ddd5e46fa3aaf2a031dd66fc4e0d8/fs/config/config.go#L249-L257
> 
> checks `os.Args`, I guess I will use a similar method, unless you know a better way

Checking os.Args is annoying, but it is the only way if you want to use the command line before the command line has initialized.

One way around this is to use `sync.Once` to run your intiialization on first call which is cheap and gets round the circular initialization problem.

**Comment by EduardDurech** [comment]
> debug with printf kind of guy!
The universal way 😆

Changes are in the pull request, just have to write documentation and unit tests.

So far I have tested it on Windows and Termux/Android. I will write a Wiki once I finish with Termux-Keystore on how this can be integrated

**Comment by animosity22** [comment]
@EduardDurech  - good to close this out and want to just comment on the PR?

**Comment by EduardDurech** [comment]
> @EduardDurech - good to close this out and want to just comment on the PR?

@animosity22 no problem, https://github.com/rclone/rclone/pull/6482?

**Comment by EduardDurech** [comment]
@animosity22 I don't see any comments on the PR, but it is completed (just needs review and for someone to "test" the tests)

**Comment by EduardDurech** [comment]
@animosity22 @ncw Sorry to bother, I just don't know if anything needs to be done on my end? Thanks!

**Comment by animosity22** [comment]
It's all with ncw as patience is the key.
